azurelinuxagent/common/protocol/wire.py

```diff
@@ -871,40 +888,53 @@ def get_vm_settings():
+                    self._vm_settings_error_reporter.report_error(error_description, _VmSettingsError.ClientError)
+                elif 500 <= response.status <= 599:
+                    self._vm_settings_error_reporter.report_error(error_description, _VmSettingsError.ServerError)
+                else:
+                    self._vm_settings_error_reporter.report_error(error_description)
+
                 raise ProtocolError("GET vmSettings [correlation ID: {0}]: {1} ({2})".format(correlation_id, response.status, error_description))
 
             for h in response.getheaders():
                 if h[0].lower() == 'etag':
                     response_etag = h[1]
                     break
             else:  # since the vmSettings were updated, the response must include an etag
-                raise ProtocolError("The vmSettings do no include an Etag [correlation ID: {0}]".format(correlation_id))
+                message = "The vmSettings do no include an Etag [correlation ID: {0}]".format(correlation_id)
+                self._vm_settings_error_reporter.report_error(message)
+                raise ProtocolError(message)
 
             response_content = self.decode_config(response.read())
             vm_settings = ExtensionsGoalStateFactory.create_from_vm_settings(response_etag, response_content)
 
-            # log the HostGAPlugin version when we fetch the vmSettings for the first time
-            is_first_vm_settings = self._vm_settings_goal_state is None
-            if is_first_vm_settings:
+            # log the HostGAPlugin version
+            if vm_settings.host_ga_plugin_version != self._host_plugin_version:
+                self._host_plugin_version = vm_settings.host_ga_plugin_version
                 message = "HostGAPlugin version: {0}".format(vm_settings.host_ga_plugin_version)
                 logger.info(message)
                 add_event(op=WALAEventOperation.HostPlugin, message=message, is_success=True)
 
             # Don't support HostGAPlugin versions older than 115
             if vm_settings.host_ga_plugin_version < FlexibleVersion("1.0.8.115"):
-                return None
+                raise _VmSettingsNotSupportedError()
 
             logger.info("Fetched new vmSettings [correlation ID: {0} New eTag: {1}]", correlation_id, vm_settings.etag)
 
             return vm_settings
 
         except ProtocolError:
             raise
-        except HttpError as http_error:
-            raise Exception("{0} [correlation ID: {1} eTag: {2}]".format(ustr(http_error), correlation_id, etag))
         except Exception as exception:
-            error = textutil.format_exception(exception)  # since this is a generic error, we format the exception to include the traceback
-            raise Exception("Error fetching vmSettings [correlation ID: {0} eTag: {1}]: {2}".format(correlation_id, etag, error))
+            if isinstance(exception, IOError) and "timed out" in ustr(exception):
```

Comment:
I don't like this string comparision but I couldn't find a version agnostic way of testing for this. Looks like the classing heirarchy is different for 2 and 3. I tried checking `errno`, but that was `None` on my python 3.8 install...

Any other possibilities? Have you looked deeply into this yet?

Reply:
don't like it either, but i found no way other than this check

the existing code in restapi.py handles IOError. telemetry shows that the exception is indeed an IOError and that the message is "timed out"

Reply:
from what I could find, the exception is a `socket.timeout` error, but from what I could tell, there's differences between the definition of that class in 2 vs. 3.

Reply:
According to telemetry it's an IOError (derived from OSError after 3.3, I think). These are the errors we have got for the last week

[HttpError] [HTTP Retry] GET http://168.63.129.16:32526/vmSettings -- Status Code 502
[HttpError] [HTTP Retry] GET http://168.63.129.16:32526/vmSettings -- Status Code 500
[HttpError] [HTTP Retry] GET http://168.63.129.16:32526/vmSettings -- Status Code 403
[HttpError] [HTTP Failed] GET http://168.63.129.16:32526/vmSettings -- IOError timed out
[HttpError] [HTTP Failed] GET http://168.63.129.16:32526/vmSettings -- IOError [Errno 113] No route to host
[HttpError] [HTTP Failed] GET http://168.63.129.16:32526/vmSettings -- IOError [Errno 111] Connection refused
[HttpError] [HTTP Failed] GET http://168.63.129.16:32526/vmSettings -- IOError [Errno 110] Connection timed out
[HttpError] [HTTP Failed] GET http://168.63.129.16:32526/vmSettings -- IOError [Errno 104] Connection reset by peer
[HttpError] [HTTP Failed] GET http://168.63.129.16:32526/vmSettings -- IOError [Errno 101] Network is unreachable